dali/pipeline/operator/builtin/merge.cc

```diff
@@ -0,0 +1,173 @@
+       output_sample_idx++) {
+    int input_category_idx = get_category_index(predicate, output_sample_idx);
+    auto &input = ws.template Input<Backend>(input_category_idx);
+
+    // get the index within input category and increment for the next occurrence.
+    int input_sample_idx = sample_idx_in_input[input_category_idx];
+    sample_idx_in_input[input_category_idx]++;
+
+    if (std::is_same_v<Backend, GPUBackend> || input.is_pinned() == *pinned_) {
+      output.SetSample(output_sample_idx, input, input_sample_idx);
+    } else {
+      // Pessimistic variant, we need to copy.
+      // TODO(klecki): Do one allocation, where samples that we share are 0-volumed - this might
+      // be perf optimization reducing the number of allocations to 1.
+      CopySampleToOutput(output, output_sample_idx, input, input_sample_idx, ws);
+    }
+  }
+  FinalizeCopy(ws);
+  WriteTestsDiagnostics(ws);
+}
+
+template <>
+void Merge<CPUBackend>::CopySampleToOutput(TensorList<CPUBackend> &output, int output_sample_idx,
+                                           const TensorList<CPUBackend> &input,
+                                           int input_sample_idx, Workspace &ws) {
+  auto &tp = ws.GetThreadPool();
+  tp.AddWork(
+      [&output, &input, output_sample_idx, input_sample_idx](int thread_idx) {
+        output.ResizeSample(output_sample_idx, input.shape()[input_sample_idx]);
+        output.CopySample(output_sample_idx, input, input_sample_idx, output.order());
+      },
+      volume(input.tensor_shape_span(input_sample_idx)));
+}
+
+
+template <>
+void Merge<CPUBackend>::FinalizeCopy(Workspace &ws) {
+  ws.GetThreadPool().RunAll();
+}
+
+
+template <>
+void Merge<GPUBackend>::CopySampleToOutput(TensorList<GPUBackend> &output, int output_sample_idx,
+                                           const TensorList<GPUBackend> &input,
+                                           int input_sample_idx, Workspace &ws) {
+  assert(false && "This codepath should not be executed");
+}
+
+
+template <>
+void Merge<GPUBackend>::FinalizeCopy(Workspace &ws) {
+  assert(false && "This codepath should not be executed");
+}
+
+
+template <typename Backend>
+void Merge<Backend>::RegisterTestsDiagnostics() {
+  this->RegisterDiagnostic("input_0_pinned", &in_0_pinned_);
+  this->RegisterDiagnostic("input_1_pinned", &in_1_pinned_);
+  this->RegisterDiagnostic("output_pinned", &out_pinned_);
+}
+
+
+template <typename Backend>
+void Merge<Backend>::WriteTestsDiagnostics(const Workspace &ws) {
+  in_0_pinned_ = ws.template Input<Backend>(0).is_pinned();
+  in_1_pinned_ = ws.template Input<Backend>(1).is_pinned();
+  out_pinned_ = ws.template Output<Backend>(0).is_pinned();
+}
+
+
+DALI_SCHEMA(experimental___Merge)
+    .DocStr(R"code(Merge batch based on a predicate.)code")
+    .NumInput(2)
+    .NumOutput(1)
+    .AddArg("predicate", "Boolean categorization of the inputs", DALI_BOOL, true)
```

Comment:
I understand the usage of this op in the scope of conditional execution. However, I think it wouldn't cost us much to make this a more usable operator by changing to the following:
```
#x = [A, B, C]
#y = [D, E]
#z = [F]
out = fn.merge(x, y, z, groups=[1, 0, 0, 2, 1, 0])
# out = [D, A, B, F, E, C]
```

Reply:
I'd be fine to limiting to 2 inputs for now, but I think that `predicate` might not be a good name for the argument. `from`, `partitions`, `groups`... would read a bit better, I think. Just food for thought

Reply:
As discussed, this is planned for a follow-up.